### PR TITLE
Enable rate limiter by http method

### PR DESCRIPTION
**Pull request by nirbenrey**
## Motivation

The request volume with http.method PATCH/DELTE/POST is much lower than the request volume with http.method GET.
To improve SM stability and avoid denial of service attacks we need to enable rate limiting by http.method

## Approach

inhance the current implementation of sm rate limter, and allow configuring the limter by "{rate}:{path}:{method}".

## Pull Request status

* [x] implementation
* [x] Unit tests
* [x] Integration tests

### CODE REVIEW

File: api/rate_limiter.go [hunk 1]
```diff
@@ -51,16 +52,18 @@ func parseRateLimiterConfiguration(input string) ([]RateLimiterConfiguration, er
 			return nil, createRateLimiterConfigurationSectionError(index, section, "no content, expected 'rate:path' format")
 		}
 		rateAndPath := strings.Split(section, ":")
```
Comment (evyaffe) [inline api/rate_limiter.go:54]: singleRateConfig (it's now not rateAndPath...)

File: api/rate_limiter.go [hunk 2]
```diff
@@ -21,6 +21,7 @@ func validateRateLimiterConfiguration(config string) error {
 type RateLimiterConfiguration struct {
 	rate       limiter.Rate
 	pathPrefix string
+	method     string
 }
 
 func createRateLimiterConfigurationSectionError(index int, section string, details string) error {
```
Comment (evyaffe) [inline api/rate_limiter.go:27]: we also need to change below "documentation" to match implementation

File: api/rate_limiter.go [hunk 3]
```diff
@@ -71,10 +74,17 @@ func parseRateLimiterConfiguration(input string) ([]RateLimiterConfiguration, er
 			if path.Clean(pathPrefix) != pathPrefix {
 				return nil, createRateLimiterConfigurationSectionError(index, section, "path is not clean, expected path '"+path.Clean(pathPrefix)+"'")
 			}
+			if len(rateAndPath) == 3 {
+				method = strings.ToUpper(rateAndPath[2])
+				if method != "POST" && method != "GET" && method != "PATCH" && method != "DELETE" {
```
Comment (evyaffe) [inline api/rate_limiter.go:79]: maybe http.MethodGet/http.MethodPost etc

File: api/rate_limiter.go [hunk 4]
```diff
@@ -71,10 +74,17 @@ func parseRateLimiterConfiguration(input string) ([]RateLimiterConfiguration, er
...
+					return nil, createRateLimiterConfigurationSectionError(index, section, "method '"+method+"' is not valid")
```
Comment (evyaffe) [inline api/rate_limiter.go:80]: "method '"+method+"' is not valid, allowed methods are: ..."

**Review by kerenlahav — APPROVED** [review]

**Review by kerenlahav — APPROVED** [review]

**Pull request merged**